reps/2023-03-10-batch-inference.md

```diff
@@ -0,0 +1,428 @@
+ds.map_batches(preprocessor1).map_batches(preprocessor2)
+```
+
+2. Predictors can be directly passed to `Dataset.map_batches`
+
+```python
+ds = ray.data.read_images(...)
+model = resnet50()
+predictor = TorchPredictor(model=model)
+
+# Alternatively, create from AIR Checkpoint.
+# predictor = TorchPredictor.from_checkpoint(air_checkpoint)
+
+predictions = ds.map_batches(predictor, compute=ray.data.ActorPoolStrategy(size=8))
+```
+
+Exception is raised if ActorPoolStrategy is not specified.
+
+```python
+ds = ray.data.read_images(...)
+model = resnet50()
+predictor = TorchPredictor(model=model)
+predictions = ds.map_batches(predictor) # Fails with ValueError.
+```
+
+3. `feature_columns` and `keep_columns` are added to the base `Predictor.predict` interface.
+
+```python
+def predict(self, 
+    data: DataBatchType, 
+    feature_columns: Optional[List[str]] = None,
+    keep_columns: Optional[List[str]] = None,
+    **kwargs) -> DataBatchType:
+    """Perform inference on a batch of data.
+
+    Args:
+        data: A batch of input data of type ``DataBatchType``.
+        feature_columns: List of columns in the preprocessed dataset to use for
+                prediction. Columns not specified will be dropped
+                from `data` before being passed to the predictor.
+                If None, use all columns in the preprocessed dataset.
+        keep_columns: List of columns in the preprocessed dataset to include
+            in the prediction result. This is useful for calculating final
+            accuracies/metrics on the result dataset. If None,
+            the columns in the output dataset will contain
+            just the prediction results.
+        kwargs: Arguments specific to predictor implementations. These are passed
+            directly to ``_predict_numpy`` or ``_predict_pandas``.
+
+    Returns:
+        Prediction result. The return type will be the same as the input type.
+    """
+```
+
+Users can specify these through `fn_args` and `fn_kwargs` in `map_batches`
+
+```python
+ds = ray.data.read_images(s3://images_with_labels)
+model = resnet50()
+predictor = TorchPredictor(model=model)
+predictions_with_labels = ds.map_batches(
+    predictor, 
+    compute=ray.data.ActorPoolStrategy(size=8),
+    fn_kwargs={"feature_columns": ["images"], "keep_columns": ["label"]}
+)
+assert "predictions" in predictions_with_labels.schema().names
+assert "label" in predictions_with_labels.schema().names
+```
+      
+### Example: Multi-stage Inference Pipeline from pre-trained model
+
+Note that Checkpoints are not on the critical path, but can still be used.
+
+#### Before
+```python
+preprocessor1 = BatchMapper(lambda x: x+1)
+preprocessor2 = TorchVisionPreprocessor(torchvision.transforms.Crop(224, 224))
+chained = Chain([preprocessor1, preprocessor2])
+
+model = pretrained_resnet()
+checkpoint = TorchCheckpoint.from_model(model, preprocessor=chained)
+batch_predictor = BatchPredictor.from_checkpoint(checkpoint)
```

Comment:
I think we want to support a `lazy=True` flag here, right? Otherwise, this wouldn't support deferred loading and/or may be confusing for users if it defaulted to lazy loading.

Reply:
This is the old BatchPredictor code path. This doesn't need to change right?

Reply:
Ah, right, I mean to comment on the new path.

Reply:
Got it. I think we should just make lazy the default, so that users don't always need to do `lazy=True` for batch prediction.

Reply:
I'm ok with that. You had a concern earlier that this would make single-process prediction benchmarks misleading though right? Since predict() would trigger loading and that could throw off timing of scripts.

Reply:
Yeah on second thought, I don't think it's that big of a concern, and will just make the API more confusing.

Reply:
removed the `lazy` arg and made lazy by default